### DISCUSSION TITLE
getCachedData - async compatibility

### DISCUSSION THREAD

**Discussion by Martichou** [nuxt/nuxt#29681] [Questions]
Hi all, I was wondering if there was any reason why getCachedData is sync and not async?

My use case is: I'd like to access IndexedDB from the getCachedData to fetch cached data but that API (of IndexedDB) is async only. So I'm ending up with an issue here trying to use async API inside a sync context.

Does any of you have an idea of what I can do to workaround that?
And is there a reason why getCachedData is sync only and don't support both async/sync? :)

Thanks for your time!

**Comment by oneminch** [comment]
Why `getCachedData` is synchronous has to do with the initial hydration process or SSR, where synchronous operations are preferred probably for performance. 

As a workaround, you can implement your own caching mechanism using a custom composable (e.g. useIndexedDB) and `useAsyncData`. You can even use similar composables from the VueUse library. The code below isn't tested. It's just to give you an idea. 

```vue
<template>
  <div>
    <div v-if="error">Error: {{ error.message }}</div>
    <div v-else-if="pending">Loading...</div>
    <div v-else>Data: {{ data }}</div>
  </div>
</template>

<script setup>
import { useIndexedDB } from '~/composables/useIndexedDB';

const { getFromIndexedDB } = useIndexedDB();

const { data, pending, error } = await useAsyncData('cachedData', async () => {
  // Fetch data from IndexedDB
  const cachedData = await getFromIndexedDB('yourKey');

  if (cachedData) {
    return cachedData;
  }

  // Fetch data if not in cache
  const response = await $fetch('/api/your-endpoint');
  const freshData = await response.json();

  // Optionally, cache the fresh data in IndexedDB here

  return freshData;
});
</script>
```
